tap_snowflake/__init__.py

```diff
@@ -422,10 +429,107 @@ def write_schema_message(catalog_entry, bookmark_properties=None):
+#                 cur.execute(list_sql)
+#                 files_in_user_stage = cur.fetchall()
+                
+#                 # 3. download files to local working directory (./<prefix>/) and move files to s3. use the same prefix for local working directory
+#                 s3_client = boto3.client('s3')                
+#                 local_working_dir = common.make_directory(prefix)
+                
+#                 LOGGER.info(f'Moving files to s3')
+#                 for file_info in files_in_user_stage:
+#                     # filename is the filename in snowflake user stage, which includes folder path in the filename. since we created 
+#                     # local working directory using the same prefix used in snowflake's user stage, we can reuse the filename to clean up 
+#                     # local working directory
+#                     filename = file_info[0]
+#                     get_sql = f"GET @~/{filename} file://{local_working_dir}/"
+#                     cur.execute(get_sql)
+
+#                     common.upload_file_to_s3(s3_client, filename, temp_s3_upload_folder)
+#                     common.remove_file(filename)
+                    
+#                     # remove processed file from snowflake's user stage
+#                     remove_file_sql = f"REMOVE @~/{filename}"
+#                     cur.execute(remove_file_sql)
+                
+#                 # log import result
+#                 total_bytes_processed = 0
+#                 total_rows_processed = 0
+#                 total_files_processed = len(copy_results)
+#                 for res in copy_results:
+#                     total_bytes_processed += res[1]
+#                     total_rows_processed += res[2]
+                
+#                 LOGGER.info(f'{catalog_entry.stream} import using internal unload finished: {total_rows_processed} rows were imported as {total_files_processed} files with total of {total_bytes_processed} bytes.')
+#             finally:
+#                 try:
+#                     # clean up, remove files from snowflake user stage
+#                     remove_sql = f"REMOVE @~/{prefix}/"
+#                     LOGGER.info(f'Running REMOVE query to clean up {prefix} folder in user stage')
+#                     cur.execute(remove_sql)
+#                     LOGGER.info(f'Cleaning up local working directory {local_working_dir}')
+#                     common.remove_directory(local_working_dir)
+#                 except:
+#                     pass
+
+
+def do_sync_external_unload(snowflake_conn, catalog_entry, columns, temp_s3_creds, temp_s3_upload_folder):
+    LOGGER.info('Stream %s is using external unload', catalog_entry.stream)
+    with snowflake_conn.connect_with_backoff() as open_conn:
+        with open_conn.cursor() as cur:
+            cur.execute('ALTER SESSION SET ENABLE_UNLOAD_PHYSICAL_TYPE_OPTIMIZATION = FALSE')
```

Comment:
whats this intended for?

Reply:
When set to true, the schema of the unloaded parquet data is determined by the column values - optimizes table columns by setting smallest precision that accepts all of the values. It didn't happen on my test cases, but added this param to be safe.